File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */

 /*
 1. l = 0
    r = 0
    res = 0
    hashMap = {}
 2. iterte over r
    for r = 0 to end {
        hashMap[r] = (hashMap[r] || 0) + 1
        while ((r - l + 1) - Max(hashMap.values()) < k) {
            hashMap[l] = hashMap[l] - 1
            l++
        }
        res = Math.max(res, r - l + 1)
    }
    return res
 */
var characterReplacement = function(s, k) {
    var l = 0
    var hashMap = {}
    var res = 0

    for (var r = 0; r < s.length; r++) {
        hashMap[s[r]] = (hashMap[s[r]] || 0) + 1
        while ((r - l + 1) - Math.max(...Object.values(hashMap)) > k) {
            hashMap[s[l]] = hashMap[s[l]] - 1
            l++
        }
        res = Math.max(res, r - l + 1)
    }
    return res
};

File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.py
'''
1. l = 0, r = 0, res = 0
charMap = {}
2.
for i from 0 to end:
    charMap[s[i]] ++
    while (r - l + 1 ) - max frequence > k:
        decerement frequency of s[l]
        l += 1
    update result
    res = max(res, current window size)
3. return res
'''

class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        l = 0
        r = 0
        res = 0
        charMap = {}

        for r in range(0, len(s)):
            charMap[s[r]] = charMap.get(s[r], 0) + 1
    
            while ((r - l + 1) - max(charMap.values()) > k):
                charMap[s[l]] -= 1
                l += 1
            
            res = max(res, r - l + 1)
        
        return res